File: 2023/src/typescript/day4/p2.ts
import { promises as fs } from 'fs';

const file_path = '/home/daniel/Documents/repos/adventofcode/2023/inputs/day4.txt';

async function solution(filePath: string) {
  try {
    const data: string = await fs.readFile(filePath, 'utf-8');
    const lines: string[] = data.trim().split("\n");

    const n: number = lines.length;
    const copies: number[][] = Array.from({ length: n }, () => []);

    for (let i = 0; i < n; i++) {
      const line: string = lines[i];
      const parts: string[] = line.split(/\s+/);
      const idx: number = parts.indexOf("|");

      const winning: number[] = parts.slice(2, idx).map(Number);
      const ours: number[] = parts.slice(idx + 1).map(Number);

      let score: number = 0;
      for (const num of ours) {
        if (winning.includes(num)) {
          score += 1;
        }
      }

      for (let j = i + 1; j <= i + score; j++) {
        copies[i].push(j);
      }
    }

    const score: number[] = Array.from({ length: n }, () => 1);

    for (let i = n - 1; i >= 0; i--) {
      for (const j of copies[i]) {
        score[i] += score[j];
      }
    }

    const totalScore: number = score.reduce((acc, curr) => acc + curr, 0);
    console.log(copies);
  } catch (error) {
    console.error(error);
  }
}

solution(file_path);

File: 2023/src/python/day3/p2.py
file_path = '/home/daniel/Documents/repos/adventofcode/2023/inputs/day3.txt'

with open(file_path, 'r') as file:
  data = file.read()
  engine_map = data.strip().split('\n')

m = len(engine_map) # Number of rows
n = len(engine_map[0]) # Number of columns

goods = [[[] for _ in range(n)] for _ in range(m)]


def is_symbol(i, j, num):
  if not(0 <= i < m and 0 <= j < n):
    return False
  
  if engine_map[i][j] == '*':
    goods[i][j].append(num)
  return engine_map[i][j] != '.' and not engine_map[i][j].isdigit()

ans = 0

for i, line in enumerate(engine_map):
  start = 0
  j = 0
  while j < n:
    start = j
    num = ''
    while j < n and line[j].isdigit():
      num += line[j]
      j += 1

    if num == '':
      j += 1
      continue

    num = int(num)

    is_symbol(i, start - 1, num) or is_symbol(i, j, num)
      
    for k in range(start-1, j+1):
      is_symbol(i - 1, k, num) or is_symbol(i + 1, k, num)

for i in range(m):
  for j in range(n):
    nums = goods[i][j]
    if engine_map[i][j] == "*" and len(nums) == 2:
      ans += nums[0] * nums[1]
        

print(ans)

